frontend/apps/reader/modules/readerbookmark.lua

```diff
@@ -196,248 +219,359 @@ function ReaderBookmark:addToMainMenu(menu_items)
+            chapter = nil
+        else
+            -- @translators In which chapter title (%1) a note is found.
+            text = T(_("in %1"), chapter)
         end
-        return a_page > b_page
+        self:addItem({
+            page = pn_or_xp,
+            text = text,
+            chapter = chapter,
+        })
     end
 end
 
+function ReaderBookmark:setDogearVisibility(pn_or_xp)
+    local visible = self:isPageBookmarked(pn_or_xp)
+    self.view.dogear:onSetDogearVisibility(visible)
+end
+
+function ReaderBookmark:isPageBookmarked(pn_or_xp)
+    local page = pn_or_xp or self:getCurrentPageNumber()
+    return self:getDogearBookmarkIndex(page) and true or false
+end
+
 function ReaderBookmark:isBookmarkInPageOrder(a, b)
     local a_page = self:getBookmarkPageNumber(a)
     local b_page = self:getBookmarkPageNumber(b)
-    if a_page == b_page then -- have bookmarks before highlights
-        return a.highlighted
+    if a_page == b_page then -- have page bookmarks before highlights
+        return not a.drawer
+    end
+    return a_page < b_page
+end
+
+function ReaderBookmark:getDogearBookmarkIndex(pn_or_xp)
+    local doesMatch
+    if self.ui.paging then
+        doesMatch = function(p1, p2)
+            return p1 == p2
+        end
+    else
+        doesMatch = function(p1, p2)
+            return self.ui.document:getPageFromXPointer(p1) == self.ui.document:getPageFromXPointer(p2)
+        end
+    end
+    local _middle
+    local _start, _end = 1, #self.ui.annotation.annotations
+    while _start <= _end do
+        _middle = bit.rshift(_start + _end, 1)
+        local v = self.ui.annotation.annotations[_middle]
+        if not v.drawer and doesMatch(v.page, pn_or_xp) then
+            return _middle
+        elseif self:isBookmarkInPageOrder({page = pn_or_xp}, v) then
+            _end = _middle - 1
+        else
+            _start = _middle + 1
+        end
     end
-    return a_page > b_page
 end
 
-function ReaderBookmark:isBookmarkInReversePageOrder(a, b)
-    -- The way this is used (by getNextBookmarkedPage(), iterating bookmarks
-    -- in reverse order), we want to skip highlights, but also the current
-    -- page: so we do not do any "a.page == b.page" check (not even with
-    -- a reverse logic than the one from above function).
-    return self:getBookmarkPageNumber(a) < self:getBookmarkPageNumber(b)
+-- add, remove, update bookmark
+
+function ReaderBookmark:addItem(item)
+    local index = self.ui.annotation:addItem(item)
+    self.ui:handleEvent(Event:new("BookmarkAdded", item))
+    self.view.footer:onUpdateFooter(self.view.footer_visible)
+    return index
 end
 
-function ReaderBookmark:isBookmarkPageInPageOrder(a, b)
-    return a > self:getBookmarkPageNumber(b)
+function ReaderBookmark:removeItem(item)
+    local index = self.ui.annotation:getItemIndex(item)
+    if item.pos0 then
+        self.ui.highlight:deleteHighlight(index) -- will call ReaderBookmark:removeItemByIndex()
```

Comment:
Do we still need ReaderBookmark to call ReaderHighlight stuff ? Shouldn't this ReaderAnnotation make all this cleaner? (Or is it still work in progress? :))

(I'm giving up a bit on my rewiew from here on :/)